base/asyncmap.jl

```diff
@@ -72,14 +73,32 @@ function next(itr::AsyncCollector, state::AsyncCollectorState)
 
     # Execute function call and save result asynchronously
     @async begin
-        itr.results[i] = itr.f(args...)
-        state.active_count -= 1
-        notify(state.task_done, nothing)
+        try
+            itr.results[i] = itr.f(args...)
+        catch e
+            try
+                if isa(itr.on_error, Function)
+                    itr.results[i] = itr.on_error(e)
+                else
+                    rethrow(e)
+                end
+            catch e2
+                state.in_error = true
+                notify(state.task_done, e2; error=true)
+
+                # The "notify" above raises an exception if "next" is waiting for tasks to finish.
+                # If the calling task is waiting on sync_end(), the rethrow() below will be captured
+                # by it.
+                rethrow(e2)
+            end
+        finally
+            state.active_count -= 1
+            notify(state.task_done, nothing)
+        end
```

Comment:
I'll incorporate the change to differentiate between done and error states and document the reasoning a bit more - that should help.